### ISSUE TITLE
exposing process.env.NODE_ENV to webpack

### ISSUE THREAD

**Issue by bochen2014** [webpack/webpack#7268]
in both `webpack@3.x` and `webpack@2.x` we config node env in `package.json` scripts 
```json
 "scripts":{
      "build":"export NODE_ENV=production&& webpack"
  }
```
and in `webpack.configure.js` we do
```js
 plugins: [
    new webpack.DefinePlugin({
      'process.env': {
        'NODE_ENV': JSON.stringify(process.env.NODE_ENV),
      },
    }),
```
to have NODE_ENV exposed to  `webpack`;

in `webpack@4.x` it seems that we still need to 
```sh
mode : process.env.NODE_ENV === 'production' ? 'production': 'development'
```

but I was wondering if this mapping is even needed. To be specific, in source code, 

https://github.com/webpack/webpack/blob/eda273d57954c1c80614a037f91b0d25fc57ab1a/lib/WebpackOptionsDefaulter.js#L299

if we change it to ` options => options.mode || process.env.NODE_ENV || "production" ` , would it bring convenience to developers (as they don't have to the mapping), or would it bring confusions?

actually, the name `options.optimization.nodeEnv` is quite obvious that it should be populated with `process.env.NODE_ENV` . why can't `webpack` do it directly?

**Comment by evilebottnawi** [comment]
@bochen2014 what about if i have `stage` NODE_ENV?

**Comment by bochen2014** [comment]
I think the convention is  anything difference from 'production' is treated as 'development'

**Comment by evilebottnawi** [comment]
@bochen2014 i disagree, i have many NODE_ENV with difference configuration some of them in development mode some on in production.

**Comment by evilebottnawi** [comment]
Also it it breaking changes, i can't change in `webpack@4`, only in `webpack@5` lso we need discussion

**Comment by bochen2014** [comment]
I don't think  `process.env.NODE_ENV` 's value matters, as `webpack` only forwards that value to modules; it's up to modules to interpret and action on;

the risk is how `definePlugin` get its `process.env.NODE_ENV` value populated. If we can always assume that it comes from environment variable 'NODE_ENV',  then I think it's safe to bypass the mapping and read environment variable directly.

I can't think of any other source where `definePlugin` could use to populate that value. 
Hope it makes sense to you

**Comment by ooflorent** [comment]
@bochen2014 In `prop-types` is it shrink at compile-time.

If you want to set a custom env you could also do this:

```js
optimization: {
  nodeEnv: "staging",
}
```

**Comment by ooflorent** [comment]
I’m against having `process.env.NODE_ENV` in my build set to the current environment. That’s a completely different thing. Even more of you have build servers.

As said in my previous comment, setting `optimization.nodeEnv` is the way to go.

**Comment by TheLarkInn** [comment]
I would agree to this. @sokra what are your thoughts?

**Comment by evilebottnawi** [comment]
But `optimization.nodeEnv` is real misleading

**Comment by michael-ciniawsky** [comment]
`process.env.NODE_ENV` is simply often used for switching the 'mode' of your/a node application between  `development` && `production` similar to how it's the intend for `config.mode` in `webpack` with the difference that it is 'accessible'  for other tools aswell (or custom `loaders` && `plugins` in `webpack` if you will) and has broader support as an defacto 'standard'. What particular `env` var doesn't matter it could also be `process.env.MODE` or even `process.env.WEBPACK_MODE`, so other tools can incooperate it. 

An example where `config.mode` is falling short atm would be `postcss-loader` and `postcss.config.js`. It would be trivial to change the `env` within `postcss-load-config` to e.g 

```js
const env = process.env.WEBPACK_MODE || process.env.NODE_ENV || 'development'
```

so 

**postcss.config.js**
```js
module.exports = (file, options, env) => ({
   console.log(env) // WEBPACK_MODE (if used with `postcss-loader`) || NODE_ENV
})
```

'just works' within the scope of supported build tools (`grunt`, `gulp`, `postcss-cli`, ...) other then `webpack` **without** requiring any special case handling from users for trivia

Also the current 'solution'  to set && use `mode` I often see is the following

**webpack.config.js**
```js
const env = process.env.NODE_ENV || 'development'

const config = { 
  mode: env // redundant
 // When using postcss-loader and 
 // one wants to rely on `mode` within the config
  module: [
    {
       test: /\.css$/,
       use: [
         { 
           loader: 'postcss-loader',
           options: { 
             config: { 
               ctx: { 
                 mode: config.mode // awkward && verbose for the exact same purpose as `NODE_ENV`
               } 
            }
         }
      ]
    }
  ]
}
```

For other `ENV` related things, `process.env` is always at your disposal e.g `process.env.BUILD_STAGE = 'staging'` etc etc... I don't really understand why folks insist `process.env.NODE_ENV` should be used here, especially if other libs/tools relying on/expecting it being either set to `development` or `production` anyways... (e.g `react`)

`optimization.nodeEnv` I don't know 💯 what is exactly does, but in case it's mainly setting `NODE_ENV` and replacing it isn't an optimization per se...(depends in the actual value). Also replacing `process.env.xyz` with it's value in code could/should simply be a default (?)

**Comment by bochen2014** [comment]
>setting optimization.nodeEnv is the way to go
I dont quite get it. Where does staging coming from in your example? Environment variable named  ‘NODE_ENV’ ? Then why can’t webpack map it itself? Why developers need to map it.

I see what @michael was saying, I.e. it doesn’t need to be an exact name making, e.g you can use WEBPACK_MODE mapped to your nodeEnv, which makes sense to me, but I haven’t run into scenarios like that.

**Comment by sokra** [comment]
Here is my opinion on that topic. I'm in general against using environment variables to configure stuff. You may ask why. I think it's a very sneaky/leaky way to pass values. They are like globals, but even worse. You can read them anywere and set them nearly anywere. Setting them is not even restricted to this process. They can be set on OS-level or in process in between, i. e. by the shell or CI server, task runner, any npm package imported somewere. Just anywere. Regarding reading them: Any npm module could use any env variable to change it's behavior. I. e. a loader could change behavior because any env variable is set to something.

In my opinion this could lead to weird "works for me (not)" behavior. Or it works locally but behaves differently on CI.

---

Instead use the methods intended for value passing. For CLIs use arguments. For code: use arguments in function calls or constructors. Best avoid static/global values and explicitly pass values.

That's also how the webpack and webpack-cli is designed:
* The CLI takes custom arguments via `--env`, which is passed to the webpack.config.js via argument (when exporting a function). You can see on the command line which values affect the process.
* The configuration file explicitly uses this value to pass it to configuration options. By reading the configuration file you see the dependencies.
* Loaders are only affected by the options passed to them. We even removed access to the webpack options from the loader API, because this pattern was violated by some loaders.

TL;DR: Avoid env variables.

PS: Some users prefer using env variable to affect the webpack.config.js. While I don't recommend that, we not actively trying to prevent users from using that.

**Comment by michael-ciniawsky** [comment]
Maybe better to leave `env` variables aside as this is imho missing the point :) (You don't mandatorily need them, it's just a common convention). What people want is a simple and 'reliable' way to switch between `development` && `production` mode. `mode` already provides that... but currently only for some `webpack` 'internals'. The moment one uses a third-party loader/plugin, he/she often needs to **additionally** fallback to e.g `process.env.NODE_ENV` to mimic `mode` (or even worst needs to 'learn'  the way the loader/plugin author choosed to handle it, e.g `options.enable || options.disable` + likely setting this option via `process.env.NODE_ENV || config.mode` somehow in the end) 🤢), this makes `mode` look like kind of redundant atm, where the reflex is 'Why not just instead using e.g `process.env.NODE_ENV` for webpack 'internals' aswell' ? There ideally should be **one** way to handle `'development' || 'production'` in `webpack` with the only ambiguity how/where to get the value of `mode` from (which isn't `webpack`'s concern). 

**webpack.config.js**
```js
import Plugin from './Plugin.js'

// If prefered (optional)
const mode = process.env.NODE_ENV || 'development'

// Using a {Function} should perhaps be 
// encouraged/documented as the 'ideal' way
// to setup the config
export default (env, argv) => {
   mode: argv.mode || mode,
   module: {
     rules: [
        {
           test: /\.ext$/,
           use: 'loader'
        }
     ]
   },
   plugins: [
     new Plugin()
   ]
}
```
### `Loaders`

```js
function loader (src, map, meta) {
   const ctx = { 
     file: this.resourcePath,
     mode: this.mode 
   }

   const config = rc(ctx.file, ctx)
   ...
}
```

### `Plugins`

#### `per plugin`

```js
class Plugin {
   constructor (options = {}) {
      this.plugin = { name: 'MyPlugin' }
      this.options = options
   }

   static mode () {
      return 'production' 
   }

   apply (compiler) {
      ...
      hook.tap(this.plugin, cb)
      ...
   }
}
```

#### `per hook`

```js
class Plugin {
  constructor (options = {}) {
    this.options = options
  }
  
  apply (compiler) {
    ...
    hook.tap({ name: 'MyPlugin' }, cb)
    hook.tap({ name: 'MyPlugin', mode: 'production' }, cb)
    hook.tap({ name: 'MyPlugin', mode: 'development' }, cb)
    ...
  }
}
```
